### ISSUE TITLE
Error when no peaks are found

### ISSUE THREAD

**Issue by ktrns** [nf-core/chipseq#128]
Hi there! 

This is an exception that you might want to catch: the experiment of the client failed and we don't see much of an enrichment, and hence no peaks are found at all. Your pipeline crashes with "Error executing process > 'PeakQC'" (plot_macs_qc.r, plot_homer_annotatepeaks.r, cat peak_annotation_header.txt macs_annotatePeaks.summary.txt > macs_annotatePeaks.summary_mqc.tsv)

```
  Error in read.table(PeakFiles[idx], sep = "\t", header = FALSE) :
    no lines available in input
  Execution halted
```

Would you be able to fix this so that the pipeline can finish off even if no peaks are found? 

Best wishes
Katrin

**Comment by drpatelh** [comment]
Hi @ktrns ! Yes, this is likely related to https://github.com/nf-core/chipseq/issues/35 :sweat: It will take quite a bit of refactoring to ignore this error and resume for the rest of the samples mainly because of the way the experimental design is incorporated into the pipeline for differential binding analysis etc. 

For now, the simplest fix is to just remove that sample from your design file and re-run. Sorry.

Would you mind sending me the output for the number of lines in each of the peak files. So if you called broad peaks then it would be something like:
`wc <OUTPUT_DIR>/results/bwa/mergedLibrary/macs/broadPeak/*.broadPeak`

or for narrow peaks:
`wc <OUTPUT_DIR>/results/bwa/mergedLibrary/macs/narrowPeak/*.narrowPeak`

**Comment by ktrns** [comment]
Hi @drpatelh, 

Sure: 0 for both non-control samples. 
bwa/mergedLibrary/Macs/narrowPeak

So if I remove my zero-peak samples, there is nothing left. But never mind, for now I can provide the atacseq pipeline results! I was just curious to also run the chipseq pipeline. 

Best wishes and many thanks
Katrin

**Comment by drpatelh** [comment]
Ah, so what did you provide as a control to the chipseq pipeline then? I'm assuming this is Cut N Run data?

**Comment by ivokwee** [comment]
I have the same problem. Is it maybe possible to skip this 'peakQC' step using some run options variable?

**Comment by drpatelh** [comment]
Quite busy at the moment with the [`viralrecon`](https://github.com/nf-core/viralrecon) pipeline but I can come back to this hopefully next week to see if I can do something in the next release which isnt far off now 👍 

However, it would be great if you can share a minimal dataset with me to be able to reproduce this error @ivokwee @ktrns? Just an IP and control will be enough 🙂

**Comment by ivokwee** [comment]
Can you at least add an option to skip this step? I think it should be in
the list of skip QA/QC steps.

**Comment by drpatelh** [comment]
Yes, I was planning to do that already but that isn't going to solve this problem because the pipeline will just fail at a later step as there are no peaks called. This is why I would need an appropriate dataset to test this on.

In the meantime, you can just remove the sample from your design file and re-run :+1:

**Comment by ivokwee** [comment]
Well, I cloned the git code, disabled the peakQC module, and re-ran. 

Here are some example files to test:
https://www.dropbox.com/s/8z0q68yajhxoqa8/nftest.tgz?dl=0ll

Also, the error is in read.table (it doesn't like an empty file). I guess we need to make dummy file with zero length read segments or something. or wrap each read with try().

Ivo

**Comment by drpatelh** [comment]
Great! Thanks for the really nice dataset to reproduce the error @ivokwee 😎 Saved me a bit of work. I have added the ability to [`--skip_peak_qc`](https://github.com/nf-core/chipseq/pull/158/commits/3a7d70b147338a464f79e2e77b14512bf3411371) and [`--skip_peak_annotation`](https://github.com/nf-core/chipseq/pull/158/commits/814d4b57e89091974472c06f7b90d16e544455f1).

But as I mentioned this is a more deep-rooted issue in the way that the pipeline is currently written. Basically, variables like `replicatesExist` and `multipleGroups` may be affected by the fact that some samples drop-out due to having 0 peaks. This would need to be taking into account immediately after the MACS2 processes for all samples have completed. Will need to revisit this at some point.

**Comment by j-andrews7** [comment]
Another bump for this thread to say this has been a frequent issue for our group, as we frequently FLAG-tag proteins and do FLAG ChIPs but include samples with no FLAG proteins as controls for the antibody.

**Comment by JoseEspinosa** [comment]
@drpatelh proposed on slack the following issue that if the files have 0 bytes we may be able to do something like [this](https://github.com/nf-core/viralrecon/blob/42e38bba3da80484bc480f53d2130642a349e8c6/subworkflows/local/assembly_spades.nf#L66)

**Comment by JoseEspinosa** [comment]
Should have been fixed in #268 I tested and works in my hands but will be nice if someone else could give it a try. I will close the issue by now.
